Design/y_e_data_storage_modes.md

```diff
@@ -0,0 +1,151 @@
+# Y and E data storage modes
+
+## Uncertainties
+
+### Status
+
+- Typically, `E` stores the standard deviation.
+- For computations such as adding data, the variance is required, thus `E` is squared and converted back to standard deviation by taking the `sqrt` at a later points.
+  Often this is done in-place, i.e., `E` temporarily stores the variance.
+- In a few cases `E` is kept as a variance until just before the end of an algorithm for performance reasons.
+  For example, `DiffractionFocussing2` needs to summing many spectra.
+  To avoid repeated evaluation of the expensive `sqrt` function for the target spectrum, all spectra are added without converting back to standard deviation after every addition. `DiffractionFocusing2.cpp` calls `Kernel::VectorHelper::rebinHistogram()` where calculations between intput and output E `vector` are performed in terms of variance. This also happens in `Kernel::VectorHelper::rebin()` 
+
+### Discussion
+
+For `E` data, the modification of the storage mode is typically very local and a fix should be rather easy.
+
+- `Histogram` currently provides functions for obtaining variances as well as standard deviations, e.g., `countVariances()` and `countStandardDeviations()`.
+  If the internal `E` vector were converted to variances by external means (as it is done currently), these access methods would do the wrong thing.
```

Comment:
In that case full protection enforced at the APIL level will not be possible. 

Something that the standard committee did when introducing the more modern c++ usages, is to make the dangerous  features undesirable and deliberately ugly looking. Take for example a const cast: `const_cast<int&>(i)` or `reinterpret_cast` compared to the c-style casts. I'm wondering if the method for the mutable access ought to look more dangerous than `mutableE()`. Such as `unsafeE()`?